File: backend/Minimarket.Api/Services/SupplierService.cs
using Minimarket.Api.DTOs;
using Minimarket.Api.Mapping;
using Minimarket.Api.Models;
using Minimarket.Api.Repositories;

namespace Minimarket.Api.Services;

public class SupplierService(ISupplierRepository supplierRepository) : ISupplierService
{
    public async Task<IReadOnlyCollection<SupplierDto>> GetAllAsync() =>
        (await supplierRepository.GetAllAsync()).Select(x => x.ToDto()).ToList();

    public async Task<SupplierDto?> GetByIdAsync(int id) =>
        (await supplierRepository.GetByIdAsync(id))?.ToDto();

    public async Task<(bool Success, string? Error, SupplierDto? Supplier)> CreateAsync(SaveSupplierDto dto)
    {
        if (string.IsNullOrWhiteSpace(dto.Name))
        {
            return (false, "El nombre del proveedor es obligatorio.", null);
        }

        var supplier = new Supplier
        {
            Name = dto.Name.Trim(),
            DocumentNumber = NormalizeOptional(dto.DocumentNumber),
            ContactName = NormalizeOptional(dto.ContactName),
            Phone = NormalizeOptional(dto.Phone),
            Email = NormalizeOptional(dto.Email),
            Address = NormalizeOptional(dto.Address),
            Notes = NormalizeOptional(dto.Notes),
            IsActive = dto.IsActive
        };

        await supplierRepository.AddAsync(supplier);
        await supplierRepository.SaveChangesAsync();

        var created = await supplierRepository.GetByIdAsync(supplier.Id);
        return (true, null, created?.ToDto());
    }

    public async Task<(bool Success, string? Error, SupplierDto? Supplier)> UpdateAsync(int id, SaveSupplierDto dto)
    {
        var supplier = await supplierRepository.GetByIdAsync(id);
        if (supplier is null)
        {
            return (false, "Proveedor no encontrado.", null);
        }

        if (string.IsNullOrWhiteSpace(dto.Name))
        {
            return (false, "El nombre del proveedor es obligatorio.", null);
        }

        supplier.Name = dto.Name.Trim();
        supplier.DocumentNumber = NormalizeOptional(dto.DocumentNumber);
        supplier.ContactName = NormalizeOptional(dto.ContactName);
        supplier.Phone = NormalizeOptional(dto.Phone);
        supplier.Email = NormalizeOptional(dto.Email);
        supplier.Address = NormalizeOptional(dto.Address);
        supplier.Notes = NormalizeOptional(dto.Notes);
        supplier.IsActive = dto.IsActive;

        supplierRepository.Update(supplier);
        await supplierRepository.SaveChangesAsync();

        var updated = await supplierRepository.GetByIdAsync(id);
        return (true, null, updated?.ToDto());
    }

    public async Task<(bool Success, string? Error)> DeleteAsync(int id)
    {
        var supplier = await supplierRepository.GetByIdAsync(id);
        if (supplier is null)
        {
            return (false, "Proveedor no encontrado.");
        }

        supplierRepository.Remove(supplier);
        await supplierRepository.SaveChangesAsync();
        return (true, null);
    }

    private static string? NormalizeOptional(string? value)
    {
        var trimmed = value?.Trim();
        return string.IsNullOrWhiteSpace(trimmed) ? null : trimmed;
    }
}

File: backend-java/minimarket-api-java/src/main/java/com/minimarket/api/service/SupplierService.java
package com.minimarket.api.service;

import com.minimarket.api.dto.SaveSupplierDto;
import com.minimarket.api.dto.SupplierDto;
import com.minimarket.api.entity.Supplier;
import com.minimarket.api.repository.SupplierRepository;
import com.minimarket.api.util.DtoMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class SupplierService {

    private final SupplierRepository supplierRepository;

    public SupplierService(SupplierRepository supplierRepository) {
        this.supplierRepository = supplierRepository;
    }

    public List<SupplierDto> getAll() {
        return supplierRepository.findAllByOrderByNameAsc()
            .stream()
            .map(DtoMapper::toDto)
            .toList();
    }

    public SupplierDto getById(Integer id) {
        return supplierRepository.findById(id)
            .map(DtoMapper::toDto)
            .orElse(null);
    }

    @Transactional
    public ServiceResult<SupplierDto> create(SaveSupplierDto dto) {
        if (dto.name() == null || dto.name().trim().isBlank()) {
            return ServiceResult.failure("El nombre del proveedor es obligatorio.");
        }

        var supplier = new Supplier();
        supplier.setName(dto.name().trim());
        supplier.setDocumentNumber(normalizeOptional(dto.documentNumber()));
        supplier.setContactName(normalizeOptional(dto.contactName()));
        supplier.setPhone(normalizeOptional(dto.phone()));
        supplier.setEmail(normalizeOptional(dto.email()));
        supplier.setAddress(normalizeOptional(dto.address()));
        supplier.setNotes(normalizeOptional(dto.notes()));
        supplier.setIsActive(Boolean.TRUE.equals(dto.isActive()));

        var saved = supplierRepository.save(supplier);
        return ServiceResult.success(DtoMapper.toDto(saved));
    }

    @Transactional
    public ServiceResult<SupplierDto> update(Integer id, SaveSupplierDto dto) {
        var supplier = supplierRepository.findById(id).orElse(null);
        if (supplier == null) {
            return ServiceResult.failure("Proveedor no encontrado.");
        }

        if (dto.name() == null || dto.name().trim().isBlank()) {
            return ServiceResult.failure("El nombre del proveedor es obligatorio.");
        }

        supplier.setName(dto.name().trim());
        supplier.setDocumentNumber(normalizeOptional(dto.documentNumber()));
        supplier.setContactName(normalizeOptional(dto.contactName()));
        supplier.setPhone(normalizeOptional(dto.phone()));
        supplier.setEmail(normalizeOptional(dto.email()));
        supplier.setAddress(normalizeOptional(dto.address()));
        supplier.setNotes(normalizeOptional(dto.notes()));
        supplier.setIsActive(Boolean.TRUE.equals(dto.isActive()));

        var updated = supplierRepository.save(supplier);
        return ServiceResult.success(DtoMapper.toDto(updated));
    }

    @Transactional
    public ServiceResult<Void> delete(Integer id) {
        var supplier = supplierRepository.findById(id).orElse(null);
        if (supplier == null) {
            return ServiceResult.failure("Proveedor no encontrado.");
        }

        supplierRepository.delete(supplier);
        return ServiceResult.success(null);
    }

    private String normalizeOptional(String value) {
        if (value == null) {
            return null;
        }

        var trimmed = value.trim();
        return trimmed.isBlank() ? null : trimmed;
    }
}
